skipchain/struct.go

```diff
@@ -405,18 +450,64 @@ func (sbm *SkipBlockMap) Store(sb *SkipBlock) SkipBlockID {
 	return sb.Hash
 }
 
+// Store stores the given SkipBlock in the service-list
+func (db *SkipBlockDB) Store(sb *SkipBlock) SkipBlockID {
+	sbOld, err := db.get(sb.Hash)
```

Comment:
There's a kind of "use after unlock" race here, not in Go, but with respect to the transactional validity of the database. It seems to me that all the actions of Store (get, if exists, then copy forward links, otherwise store as is) need to be in one boltdb txn.

Reply:
Inside `db.get`, I make a copy of whatever is returned by the transaction. So the operations in `Store` is done on a copy of the data. That should be OK right?

Reply:
If there are two requests to add blocks that arrive on two TCP streams concurrently, then they will execute in two goroutines. If they are both writes of new skipblocks, I think they will be racy with respect to each other.